### DISCUSSION TITLE
Modbus ID second inverter

### DISCUSSION THREAD

**Discussion by xsirpreisx** [WillCodeForCats/solaredge-modbus-multi#274] [General]
Hi,

I have two inverters and one modbus meter and one BYD battery. Ever since installation the master inverter has modbus ID 1 and the second inverter has modbus ID 3. 

I don't want to change the modbus IDs, because it works with everything else that I'm connecting via modbus (using a modbus proxy that connects to my SolarEdge master). Both openWB and evcc (both EV solutions) connect to that modbus proxy and don't have a problem with the second inverter having id 3.

Somehow this integration expects the second inverter to have modbus id 2. Can this be configured somehow?

**Comment by WillCodeForCats** [comment]
You can try adding the integration multiple times for each modbus address if you don't want to use the supported scheme.

**Reply by xsirpreisx** [reply]
I would like to be able to have combined "daily production" for both invertes (the whole system in total), "daily consumption" and so on. Will I be able to have those when adding the integration multiple times? Or will I have to create sensors to sum stuff manually?

**Reply by WillCodeForCats** [reply]
Daily production and daily consumption are not part of the the modbus spec, so you would need to calculate those values manually either way. I am working on a "virtual inverter" feature that will automatically create summary data like that, but that feature is not yet available.

**Reply by xsirpreisx** [reply]
Ok, that makes sense.

I'm looking for the easiest way to implement this: https://community.home-assistant.io/t/updated-solaredge-modbus-full-setup-guide-with-energy-dashboard-integration-for-installations-with-battery-connected/340956

If I understand correctly, I will have to add this integration twice, and then I'll have to manually create sensors which will sum production, export etc, and only then I'll be able to go forward implementing what was created in that thread on the HA forum?

Or would it be like that anyway if I were able to use this integration just once (if my inverters had modbus ids 1 and 2)?

**Reply by xsirpreisx** [reply]
And you probably don't have a time line for that "virtual inverter" feature, I guess? :-)

**Reply by xsirpreisx** [reply]
I tried adding the integration twice, but I am getting the error message that "the inverter is already configured". I can only connect to the master inverter, as the second one is only connected via RS485, it's not in the network. So I'm not sure how to add the integration multiple times.

**Reply by xsirpreisx** [reply]
Sorry for spaming. I just found your configuration page with the very useful information on it (especially that my meter can stay at modbus id 2 as its a separate bus). I will try to set my slave inverter to modbus id 2 and then use your integration. Thank you :-)

**Reply by WillCodeForCats** [reply]
I don't have an ETA on the virtual inverter feature, but I can start looking at doing some pre-releases with it.

**Reply by xsirpreisx** [reply]
Thank you! I'm not sure I'd need it, so you don't have to hurry for me :-D I was able to add both inverters, meter and battery after changing modbus ids and it seems to be working fine.

I used the code from the HA forum (https://community.home-assistant.io/t/solaredge-modbus-configuration-for-single-inverter-and-battery/464084) to calculate all the needed sensors. Only problem I have (and it might be something you might have to look at when creating your "virtual inverter" feature) with that is that somehow the "panel to house" calculation while battery is charging is somehow not correct. This value is always too high while the battery is charging (and only then).

The code they use is


```
- name: "Solar Panel To House W"
        unique_id: solar_panel_to_house_w
        unit_of_measurement: "W"
        icon: mdi:solar-power
        state: >
          {% set i1_dc_power = states('sensor.solaredge_i1_dc_power') | float(0) %}
		  {% set i2_dc_power = states('sensor.solaredge_i2_dc_power') | float(0) %}																		   
          {% set i1_ac_power = states('sensor.solaredge_i1_ac_power') | float(0) %}
          {% set i2_ac_power = states('sensor.solaredge_i2_ac_power') | float(0) %}
		  {% set b1_dc_power = states('sensor.solaredge_b1_dc_power') | float(0) %}
          {% set m1_ac_power = states('sensor.solaredge_m1_ac_power') | float(0) %}
          {% set inverter_effectiveness = states('sensor.solar_inverter_effectiveness') | float(0) %}
          {% if (b1_dc_power >= 0 and m1_ac_power > 0) %}
            {% if ((i1_dc_power + i2_dc_power) < 0 and (i1_ac_power + i2_ac_power) <= 0) %}
              {{ ((i1_dc_power + i2_dc_power) - m1_ac_power) }}
            {% else %}
              {{ ((i1_ac_power + i2_ac_power) - m1_ac_power) }}
            {% endif %}
          {% elif (b1_dc_power >= 0 and m1_ac_power <= 0) %}
            {% if ((i1_dc_power + i2_dc_power) < 0 and (i1_ac_power + i2_ac_power) <= 0) %}
              {{ (i1_dc_power + i2_dc_power) }}
            {% else %}
              {{ (i1_ac_power + i2_ac_power) }}
            {% endif %}
          {% elif (b1_dc_power < 0) %}
            {% if ((i1_dc_power + i2_dc_power) + b1_dc_power < 0) %}
              0
            {% else %}
              {{ (((i1_dc_power + i2_dc_power) + b1_dc_power) * inverter_effectiveness) }}
            {% endif %}   
          {% else %}
            0
          {% endif %}
        availability: >
          {{ states('sensor.solaredge_i1_dc_power') | is_number and states('sensor.solaredge_i2_dc_power') | is_number and states('sensor.solaredge_i1_ac_power') | is_number and states('sensor.solaredge_i2_ac_power') | is_number and states('sensor.solaredge_b1_dc_power') | is_number and states('sensor.solaredge_m1_ac_power') | is_number }}
```


It seems that sometimes DC power of both inverters is summed up, and other times the AC power of both inverters is used. In my case summing up AC power results in too high "panel to grid" values. I had to change the code as follows, always only using the sum of DC power. I looked at the results for the past few days and they look much better. Might not be perfect, I'm not sure.
   
```
{% if (b1_dc_power >= 0 and m1_ac_power > 0) %}
            {{ ((i1_dc_power + i2_dc_power) - m1_ac_power) }}
          {% elif (b1_dc_power >= 0 and m1_ac_power <= 0) %}
            {{ (i1_dc_power + i2_dc_power) }}
          {% elif (b1_dc_power < 0) %}
            {% if ((i1_dc_power + i2_dc_power) + b1_dc_power < 0) %}
              0
            {% else %}
              {{ (((i1_dc_power + i2_dc_power) + b1_dc_power) * inverter_effectiveness) }}
            {% endif %}   
          {% else %}
            0
          {% endif %}
```

Another thing I'm not sure about is inverter_effectiveness (and battery_effectiveness). With my current calculation, inverter_effectiveness can get above 200 %, which doesn't make much sense from my point of view... But that hasn't created with the sensor values that I'm using, as far as I can see.

Sorry for the long post, which doesn't even really belong here.. Was just thinking that maybe it will help with your development of that part. Or maybe you know what I'm talking about and know how to calculate those things correctly :-D

Edit: Ok, it doesn't work in all situations, just adding dc power of both inverterse. Sun is slowly rising and right now I'm getting a negative value. Guess it's not that easy afterall...
